File: Practica1/Backend-python/app/services/playlist_service.py
from app.repositories.playlist_repository import (
    find_movie_by_id,
    add_to_playlist,
    get_user_playlist,
    remove_from_playlist
)
from app.repositories.notification_repository import create_notification
import psycopg2
import os



def add_movie_to_playlist(id_usuario, id_pelicula):

    movie = find_movie_by_id(id_pelicula)

    if not movie:
        raise Exception("Película no existe")

    if movie[6] != "Disponible":
        raise Exception("La película no está disponible")

    try:
        result = add_to_playlist(id_usuario, id_pelicula)

    
        create_notification(
            id_usuario,
            "Película agregada a tu lista 🎬"
        )

        return result

    except psycopg2.errors.UniqueViolation:
        raise Exception("La película ya está en tu lista")


def list_user_playlist(id_usuario):

    movies = get_user_playlist(id_usuario)
    
    bucket = os.getenv("AWS_BUCKET_NAME")
    region = os.getenv("AWS_REGION")
    base_url = f"https://{bucket}.s3.{region}.amazonaws.com"

    result = []
    for m in movies:
        result.append({
            "id_pelicula": m[0],
            "titulo": m[1],
            "director": m[2],
            "anio_estreno": m[3],
            "url_contenido": m[4],
            "poster_url": f"{base_url}/{m[5]}",
            "estado": m[6]
        })

    return result


def delete_from_playlist(id_usuario, id_pelicula):
    remove_from_playlist(id_usuario, id_pelicula)

File: Practica1/Backend-Nodejs/src/services/playlist.service.js
const playlistRepository = require("../repositories/playlist.repository");
const notificationRepository = require("../repositories/notification.repository");

async function addMovieToPlaylist(id_usuario, id_pelicula) {
  const movie = await playlistRepository.findMovieById(id_pelicula);

  if (!movie) {
    throw new Error("Película no existe");
  }

  if (movie.estado !== "Disponible") {
    throw new Error("La película no está disponible");
  }

  try {
    const result = await playlistRepository.addToPlaylist(id_usuario, id_pelicula);

    
    await notificationRepository.createNotification(
      id_usuario,
      "Película agregada a tu lista"
    );

    return result;

  } catch (error) {
    if (error.code === "23505") {
      throw new Error("La película ya está en tu lista");
    }
    throw error;
  }
}

async function listUserPlaylist(id_usuario) {
  return await playlistRepository.getUserPlaylist(id_usuario);
}

async function deleteFromPlaylist(id_usuario, id_pelicula) {
  await playlistRepository.removeFromPlaylist(id_usuario, id_pelicula);
}

module.exports = {
  addMovieToPlaylist,
  listUserPlaylist,
  deleteFromPlaylist
};